test_conformance/subgroups/subhelpers.h

```diff
@@ -56,219 +66,943 @@ class subgroupsAPI {
 {
-    static bool val(cl_device_id device) { return int64_ok(device); }
-};
-template <> struct TypeCheck<cl_float>
+    lhs = static_cast<Ty>(rhs);
+}
+
+// set scalar value to half scalar
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_scalar_type::value>::type
+set_value(Ty &lhs, const cl_ulong &rhs)
 {
-    static bool val(cl_device_id) { return true; }
-};
-template <> struct TypeCheck<cl_half>
+    lhs.data = rhs;
+}
+
+
+// compare for common vectors
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_vector_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
 {
-    static bool val(cl_device_id device)
+    const int size = sizeof(Ty) / sizeof(typename TypeManager<Ty>::scalar_type);
+    for (auto i = 0; i < size; ++i)
     {
-        return is_extension_available(device, "cl_khr_fp16");
+        if (lhs.s[i] != rhs.s[i])
+        {
+            return false;
+        }
     }
-};
-template <> struct TypeCheck<double>
+    return true;
+}
+
+
+// compare for vectors 3
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_vector_size3::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
 {
-    static bool val(cl_device_id device)
+    for (auto i = 0; i < 3; ++i)
     {
-        int error;
-        cl_device_fp_config c;
-        error = clGetDeviceInfo(device, CL_DEVICE_DOUBLE_FP_CONFIG, sizeof(c),
-                                (void *)&c, NULL);
-        if (error)
+        if (lhs.data.s[i] != rhs.data.s[i])
         {
-            log_info(
-                "clGetDeviceInfo failed with CL_DEVICE_DOUBLE_FP_CONFIG\n");
             return false;
         }
-        return c != 0;
     }
-};
+    return true;
+}
+
+// compare for half vectors
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_vector_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    const int size = sizeof(Ty) / sizeof(typename TypeManager<Ty>::scalar_type);
+    for (auto i = 0; i < size; ++i)
+    {
+        if (lhs.data.s[i] != rhs.data.s[i])
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+
+// compare for scalars
+template <typename Ty>
+typename std::enable_if<std::is_scalar<Ty>::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    return lhs == rhs;
+}
+
+// compare for scalar halfs
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_scalar_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    return lhs.data == rhs.data;
+}
+
+
+inline bool isnan_half(const subgroups::cl_half &x)
+{
+    return (x.data & 0x7fff) > 0x7c00;
+}
+
+template <typename Ty> inline bool compare_ordered(const Ty &lhs, const Ty &rhs)
+{
+    return lhs == rhs;
+}
+
+template <>
+inline bool compare_ordered(const subgroups::cl_half &lhs,
+                            const subgroups::cl_half &rhs)
+{
+    return lhs.data == rhs.data && !isnan_half(lhs);
```

Comment:
Is this safe?  Consider comparing positive zero with negative zero.

Reply:
Ben
Reading in the Internet [info ](https://en.wikipedia.org/wiki/Signed_zero)and comparing compilation VS and gcc the results is always true (equal)
https://godbolt.org/z/TMzEET

Reply:
Your example is testing float. Half being a non-standard type in C/C++, the host code probably cannot rely on the same assumption as for `float` or `double`.

Reply:
Right - in `cl_platform.h` we've defined `cl_half` to be an unsigned 16-bit integer that can encode the bit patterns for a half but this means that it does not necessarily obey the floating-point behavior since there are no overloaded operators.  I believe this demonstrates the problem: https://godbolt.org/z/d17rd6

One solution (untested) might be:

```c
return cl_half_to_float(lhs.data) == cl_half_to_float(rhs.data);
```

(This comment applies equally well to #1074, if this PR gets closed.)

Reply:
I haven't found the usages of this yet, but we should define and document the semantics we're trying to achieve. Most tests don't want to do a straight IEEE-754 comparison because NaNs will fail. Typical options are:

1. Bitwise exact (just compare as integers)
2. Bitwise exact, excluding NaNs (allows the NaN payload to vary from the reference)
3. Bitwise exact, excluding NaNs and subnormals (if flush to zero is allowed)

There are examples of #3 in the image tests (e.g. https://github.com/KhronosGroup/OpenCL-CTS/blob/24e6a9125c44ef9fe42603b605a685af9acb4bdf/test_conformance/images/kernel_read_write/test_iterations.cpp#L1173)

Reply:
Then I am not sure for now which way to go :) I will use Ben's proposal. Fixed https://github.com/KhronosGroup/OpenCL-CTS/pull/1074